### DISCUSSION TITLE
Any known reasons for differing numerical outputs after 0.3.0 update?

### DISCUSSION THREAD

**Discussion by DanPuzzuoli** [jax-ml/jax#9556] [Q&A] [answered]
Hi,

Apologies in advance for the vagueness of this question, but are there any known/expected situations in which the numerical output of "core" JAX operations should be different between 0.3.0 and the previous release? (By "core" I mean to imply that the code I'm looking at doesn't use any newer JAX functionality, its all built from simple `jax.numpy` stuff, array indexing, `vmap`, and `scan`.) 

Long story short, I have some JAX code that, after upgrading to 0.3.0, is outputting substantially different numerical values than when its run on the previous release. The problem occurs whether I'm on jaxlib 0.3.0 or 0.1.76, so I think whatever is causing this difference is coming from a difference between jax 0.2.28 -> 0.3.0. (Basically I can switch between which release is installed and observe the numerical difference, with relative errors on the order of `1e-1` for some array entries.) 

I need to keep diving in to isolate where the difference is occurring (at which point I can share more detailed info) - but at this point I wanted to check in here to see if anything like this is known and/or coming up elsewhere.

Thanks!

Dan

**Comment by hawkinsp** [comment]
Despite the version number bump, 0.3 didn't actually change much in the way of code (it's mostly a new version numbering scheme). So I think any numerical difference is unintended.

I think we'd need a repro to say more.

**Reply by DanPuzzuoli** [reply]
Okay 👍 . I'll keep digging to identify it and try to come up with something self contained to share.

**Comment by DanPuzzuoli** [comment, marked as answer]
Okay based on a suggestion by a colleague I've identified which commit in the `0.3.0` release introduces the difference (installed from source and binary searched the commit list in https://github.com/google/jax/compare/jax-v0.2.28...jax-v0.3.0 ).

The test I'm running "passes" (produces expected results which are consistent with previous commits) up to and including commit
```
e219ea08f5edf766ed9eee48f413a922bad57568
``` 
but starts failing on commit 
```
8ca6622c0b78f157c629187fa80d94f9f515078d
```

I still need to identify in exactly what part of our code this ultimately gets used.

**Reply by hawkinsp** [reply]
It's surprising to me that this leads to any change at all in numerics, so I definitely want to look into this.

The only place I'd expect to see a significant change in anything (mostly performance) would be when batching (`vmap`) conditionals.

**Reply by DanPuzzuoli** [reply]
Yeah there's a couple of layers of `vmap` being used, with one instance of `switch` that may be responsible for this - I think my characterization of this as a "numerical" issue is incorrect at this point. Should hopefully have a small example shortly.

**Reply by DanPuzzuoli** [reply]
Okay based on narrowing things down I have a small example:

```
from jax import vmap
from jax.lax import switch

f1 = lambda t: t**3
f2 = lambda t: t**2
f3 = lambda t: t
f4 = lambda t: 2 * t
f5 = lambda t: 5 * t
f6 = lambda t: 6 * t
f_list = [f1, f2, f3, f4, f5, f6]

def single_eval(idx, t):
    return switch(idx, f_list, t)

vmap(single_eval, in_axes=(0, None))(np.array([0, 1, 2, 3, 4, 5]), 0.5)
```

Output should just be an array containing `[f(0.5) for f in f_list]`. Commit `e219ea08f5e` correctly gives:
```
DeviceArray([0.125, 0.25 , 0.5  , 1.   , 2.5  , 3.   ], dtype=float64)
```

whereas commit `8ca6622c0b78` gives:
```
DeviceArray([0.125, 0.25 , 0.5  , 1.   , 3.   , 3.   ], dtype=float64)
```

It seems that as you increase the length of the function list, its the second last output that is always incorrectly overwritten to be the same as the last.

Also, what makes this hard to find, and may help indicate what the problem is, is that this only starts happening once `len(f_list) > 4`. (This explains why my unittests were passing - they weren't hittnig examples big enough to detect this, whereas the one bigger integration test I was doing did.)

**Reply by hawkinsp** [reply]
Thanks, that's perfect. I see the bug. Oops. PR coming shortly.

**Reply by hawkinsp** [reply]
https://github.com/google/jax/pull/9563 should fix.

**Reply by hawkinsp** [reply]
Now merged at head.

**Reply by DanPuzzuoli** [reply]
Awesome, thanks for handling this so quickly! I'll mark your upper answer as the "answer" to indicate this is resolved.

**Reply by DanPuzzuoli** [reply]
Actually nm, I'll mark this thread as the answer as the upper one might give the false impression that there wasn't an issue.
